new_platforms/scripts/VirtualMachineSgxSettings.psm1

```diff
@@ -0,0 +1,283 @@
+
+Function Get-VMSgx
+{
+    [CmdletBinding()]
+    Param
+    (
+        # Name of the VM to get the SGX settings for.
+        [Parameter(Mandatory=$True, ParameterSetName="ByName", Position=0)]
+        [ValidateNotNullOrEmpty()]
+        [String]$VmName,
+
+        # VM to get the SGX settings for.
+        [Parameter(Mandatory=$True, ParameterSetName="ByObject", Position=0)]
+        [ValidateNotNull()]
+        [Microsoft.HyperV.PowerShell.VirtualMachine]$Vm
+    )
+
+    Process
+    {
+        If (!$Vm)
+        {
+            $Vm = Get-VM $VmName
+        }
+
+        $Mem = Get-VMSgxMemorySettingData $Vm
+        Return [VirtualMachineSgx]::New($Mem)
+    }
+}
+
+### -----------------------------------
+### Set-VMSgx Cmdlet
+### -----------------------------------
+
+Function Set-VMSgx
+{
+    [CmdletBinding()]
+    Param
+    (
+        # Name of the VM to configure.
+        [Parameter(Mandatory=$True, ParameterSetName="ByName", Position=0)]
+        [ValidateNotNullOrEmpty()]
+        [String]$VmName,
+
+        # VM to configure.
+        [Parameter(Mandatory=$True, ParameterSetName="ByObject", Position=0)]
+        [ValidateNotNull()]
+        [Microsoft.HyperV.PowerShell.VirtualMachine]$Vm,
+
+        # Whether SGX is enabled in the VM.
+        [Parameter(Position=1)]
+        [Bool]$IsSgxEnabled = $False,
+
+        # Desired SGX EPC Memory Size for the VM (in MB).
+        [Parameter(Position=2)]
+        [UInt64]$SgxSize = 0,
+
+        # Desired default SGX Launch Control Mode for the VM.
+        [Parameter(Position=3)]
+        [String]$SgxLaunchControlDefault = [String]::Empty,
+
+        # Desired SGX Launch Control Mode for the VM.
+        [Parameter(Position=4)]
+        [UInt32]$SgxLaunchControlMode = 0,
+
+        # Ignore warnings.
+        [Parameter()]
+        [Switch]$Force = $False
+    )
+
+    Process
+    {
+        # If the user passed a VM name, retrieve the VM object.
+        If ($Vm -eq $Null)
+        {
+            $Vm = Get-VM $VmName
+        }
+
+        # If the VM is not Off, the function cannot proceed.
+        If ($Vm.State -ne [Microsoft.HyperV.PowerShell.VMState]::Off)
+        {
+            Write-Error ($Strings.ERR_VM_STATE -f $Vm.Name)
+        }
+
+        # At the time of this writing, the following conditions hold:
+        #
+        # 1. SGX virtualization is only supported on Gen2 VMs:
+        # 1.1. The Intel SGX driver is not yet merged into mainline Linux and must be compiled manually;
+        #      For VMs with Secure Boot on, this means that the module must be signed with a custom key;
+        #      Note: Only Gen2 VMs support Secure Boot.
+        # 2. Checkpoints are not supported with SGX turned on.
+
+        If (!$Force -and $IsSgxEnabled)
+        {
+            # Check VM generation.
+            If ($Vm.Generation -lt 2)
+            {
+                Write-Warning ($Strings.WARN_GEN2_TITLE -f $Vm.Name)
+                Write-Warning $Strings.WARN_GEN2_TEXT
```

Comment:
Given the above comment about SGX support, should we not abort/throw here?

Reply:
That's one of the issues I wanted your input on.

The WMI method invoked on line 247 does its own validation of the desired settings. However, it does not provide the user with any meaningful information in case of failure, other than a numeric error code. As a result, I decided to add some checking of my own. However, said checking duplicates the checking done by the WMI method, and they might, further down the road, diverge.

Hence, I decided to make the checks warn or inform rather than terminate, except when the error is obvious (i.e. enable SGX with no SGX memory, or when the VM is running).

Reply:
One way is to run the check after the failure of the WMI call. When it fails, you can test for the familiar failure patterns to help user diagnose.

Reply:
Good idea.

Reply:
I've refactored the error checking:

For conditions that won't cause the modification command to fail, I check them before trying to change anything, unless the user passes in `-Force`. Those checks are useful to give the user an idea of what might go wrong when trying to start VMs.

For conditions that are likely to have caused the modification command to have failed, I check them only after the failure is detected.

Reply:
I wonder if we are overloading -Force here. In my head, -Force is ignore warnings/possible conflicts and proceed. We are doing that even without -Force. Does it make sense to use -Force just to change warning output, without any real change in the code flow?

Reply:
Agreed; removed the usage of `Force` altogether. The user can use PowerShell's built-in warning suppression mechanisms.